lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,893 @@
+/// name; 2) number of inputs; 3) number of outputs; 4) comparison operation
+/// type (if applied); 5) whether the elastic component is for the control path
+/// (if applied).
+static std::string getSubModuleName(Operation &oldOp) {
+  std::string subModuleName = oldOp.getName().getStringRef().str() + "_" +
+                              std::to_string(oldOp.getNumOperands()) + "ins_" +
+                              std::to_string(oldOp.getNumResults()) + "outs";
+
+  if (auto comOp = dyn_cast<mlir::CmpIOp>(oldOp))
+    subModuleName += "_" + stringifyEnum(comOp.getPredicate()).str();
+
+  if (auto ctrlAttr = oldOp.getAttr("control")) {
+    if (ctrlAttr.cast<BoolAttr>().getValue())
+      subModuleName += "_ctrl";
+  }
+
+  return subModuleName;
+}
+
+//===----------------------------------------------------------------------===//
+// FIRRTL Top-module Related Functions
+//===----------------------------------------------------------------------===//
+
+/// Currently we are not considering clock and registers, thus the generated
+/// circuit is pure combinational logic. If graph cycle exists, at least one
+/// buffer is required to be inserted for breaking the cycle, which will be
+/// supported in the next patch.
+FModuleOp createTopModuleOp(handshake::FuncOp funcOp,
+                            ConversionPatternRewriter &rewriter) {
+  using ModulePort = std::pair<StringAttr, FIRRTLType>;
+  llvm::SmallVector<ModulePort, 8> ports;
+
+  // Add all inputs of funcOp.
+  int args_idx = 0;
+  for (auto &arg : funcOp.getArguments()) {
+    StringAttr portName =
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(arg.getType(), /*isFlip=*/false);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  // Add all outputs of funcOp.
+  for (auto portType : funcOp.getType().getResults()) {
+    StringAttr portName =
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(portType, /*isFlip=*/true);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  // Create a FIRRTL module, and inline the funcOp into it.
+  FModuleOp topModuleOp = rewriter.create<FModuleOp>(
+      funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()),
+      ArrayRef<ModulePort>(ports));
+  rewriter.inlineRegionBefore(funcOp.getBody(), topModuleOp.getBody(),
+                              topModuleOp.end());
+
+  // Merge the second block (inlined from funcOp) of the top-module into the
+  // entry block.
+  auto blockIterator = topModuleOp.getBody().begin();
+  Block *entryBlock = &*blockIterator;
+  Block *secondBlock = &*(++blockIterator);
+
+  // Replace uses of each argument of the second block with the corresponding
+  // argument of the entry block.
+  for (int i = 0, e = secondBlock->getNumArguments(); i < e; ++i) {
```

Comment:
nit: I think arguments are in a linked list, so this is unnecessarily N^2.  N is usually small, but can be large (e.g. O(1000)) in real world code.